lang/cs/Org.Apache.REEF.IMRU.Examples/PipelinedBroadcastReduce/PipelinedBroadcastAndReduceWithRandomFailures.cs

```diff
@@ -0,0 +1,389 @@
+                    var thread = new Thread(exit.After);
+                    maxTimeout = minTimeoutInSec + (maxTimeoutInMilliSec / 1000.0);
+                    thread.Start();
+                }
+                _taskState = (MapTaskState<int[]>)taskState;
+
+                var taskIdSplit = taskId.Split('-');
+                _retryIndex = int.Parse(taskIdSplit[taskIdSplit.Length - 1]);
+
+                if (chosenToFail)
+                {
+                    Logger.Log(Level.Info,
+                        "SenderMapFunctionFT.Ctor: TaskId: {0}, RetryNumber: {1}, Dimensions: {2}, Failure Prob: {3}, MinTimeoutInSec: {4}, MaxTimeoutInSec: {5:F2}.",
+                        _taskId,
+                        _retryIndex,
+                        dim,
+                        failProb,
+                        minTimeoutInSec,
+                        maxTimeout);
+                }
+                else
+                {
+                    Logger.Log(Level.Info,
+                        "SenderMapFunctionFT.Ctor: TaskId: {0}, RetryNumber: {1}, Dimensions: {2}, *****Not Chosen for Failure in this retry attempt*****",
+                        _taskId,
+                        _retryIndex,
+                        dim);
+                }
+            }
+
+            /// <summary>
+            /// Map function
+            /// It simply takes the input, does a verification for the task state and returns the same input value. 
+            /// </summary>
+            /// <param name="mapInput">integer array</param>
+            /// <returns>The same integer array</returns>
+            int[] IMapFunction<int[], int[]>.Map(int[] mapInput)
+            {
+                int[] previousValue = _taskState.CurrentValue;
+
+                // In this example, when task is re-started, with the task state management, it should continue from the previous number
+                if (previousValue != null && previousValue[0] > mapInput[0])
+                {
+                    throw new Exception(string.Format(CultureInfo.CurrentCulture,
+                        "The previous value received was {0} but the value received is {1} which is smaller than the previous value.",
+                        previousValue[0],
+                        mapInput[0]));
+                }
+
+                _taskState.CurrentValue = mapInput;
+
+                _iterations++;
+                Logger.Log(Level.Info, "SenderMapFunctionFT.Map: TaskId: {0} Received value {1} in iteration {2}.", _taskId, mapInput[0], _iterations);
+
+                return mapInput;
+            }
+
+            public void Dispose()
+            {
+            }
+        }
+
+        /// <summary>
+        /// The Update function for integer array broadcast and reduce
+        /// </summary>
+        internal sealed class BroadcastSenderReduceReceiverUpdateFunctionFT : IUpdateFunction<int[], int[], int[]>
```

Comment:
If this is the same as the class defined in PipelinedBroadcastAndReduceWithFaultTolerant.cs, you might consider to reuse.  That one is defined as internal.